src/oData/jsonToOData.ts

```diff
@@ -0,0 +1,217 @@
+		if (regexp.pattern == null) {
+			throw new Error(
+				`Regex object defined but the pattern property was empty`,
+			);
+		}
+		if (regexp.flags != null && regexp.flags !== 'i') {
+			throw new Error(`Regex flag ${regexp.flags} is not supported`);
+		}
+		if (regexp.flags === 'i') {
+			return {
+				$contains: [
+					{
+						$tolower: {
+							$: parentKeys.length === 1 ? parentKeys[0] : parentKeys,
+						},
+					},
+					regexp.pattern.toLowerCase(),
+				],
+			};
+		}
+		return wrapValue(parentKeys, { $contains: regexp.pattern });
+	}
+	// Tiny optimization
+	if (typeof value.not !== 'boolean' && value.not?.const != null) {
+		return wrapValue(parentKeys, { $ne: value.not.const });
+	}
+	if (value.not != null && typeof value.not !== 'boolean') {
+		// !properties && !contains && type = string | number | boolean
+		return { $not: handlePrimitiveFilter(parentKeys, value.not) };
+	}
+	const filters: any[] = [];
+	for (const [$op, jsonSchemaProps] of Object.entries(comparisonOperatorMap)) {
+		for (const jsonSchemaProp of jsonSchemaProps) {
+			if (value[jsonSchemaProp] != null) {
+				filters.push(wrapValue(parentKeys, { [$op]: value[jsonSchemaProp] }));
+			}
+		}
+	}
+	if (filters.length === 0) {
+		throw new Error('');
+	}
+	return maybePluralOp('$and', filters);
+};
+
+const wrapValue = (parentKeys: string[], value: any): PineFilterObject => {
+	for (let i = parentKeys.length - 1; i >= 0; i--) {
+		value = { [parentKeys[i]]: value };
+	}
+	return value;
+};
+
+const handleFilterArray = (
+	parentKeys: string[],
+	filterObj: JSONSchema,
+): PineFilterObject => {
+	const filters: PineFilterObject[] = [];
+	if (filterObj.minItems != null && filterObj.minItems > 1) {
+		const field = parentKeys[parentKeys.length - 1];
+		const parent$alias = parentKeys[parentKeys.length - 2]
+			? createAlias(parentKeys[parentKeys.length - 2])
+			: null;
+		filters.push({
+			$ge: [
+				wrapValue([...(parent$alias != null ? [parent$alias!] : []), field], {
+					$count: {},
+				}),
+				filterObj.minItems,
+			],
+		});
+	}
+	// Post-MVP: maxItems
+	if (filterObj.contains) {
+		const parentKey = parentKeys[parentKeys.length - 1];
+		const $alias = createAlias(parentKey);
+		const nestedFilters = filterBuilder([$alias], filterObj.contains);
+		if (nestedFilters) {
+			filters.push({
+				[parentKey]: {
+					$any: { $alias, $expr: nestedFilters },
+				},
+			});
+		}
+	}
+
+	if (filters.length === 0) {
+		filters.push({ 1: 1 });
+	}
+	return maybePluralOp('$and', filters);
+};
+
+const handleOperators = (
+	parentKeys: string[],
+	filter: JSONSchema,
+): PineFilterObject | undefined => {
+	if (!filter.anyOf && !filter.allOf) {
+		throw new Error('Calling handleOperators without anyOf and allOf');
+	}
+	const operator = filter.anyOf || filter.oneOf ? '$or' : '$and';
+	const filters = filter.anyOf || filter.oneOf || filter.allOf;
+	return filterBuilder(parentKeys, maybePluralOp(operator, filters!));
+};
+
+export const filterBuilder = (
```

Comment:
Would be great to move this file to a separate package as a Hack-Friday once we can take a breath.
Call it `jsonschema-to-odata` and export
* filterBuilder -> convertToPineClientFilter
* convertToOData (which just calls convertToPineClientFilter and then uses pineclient's `.compile`)